src/eos-updater-poll-common.c

```diff
@@ -303,6 +303,78 @@ get_booted_refspec (gchar **booted_refspec,
   return TRUE;
 }
 
+static GFile *
+get_refspec_to_upgrade_on_file_from_deployment (OstreeSysroot    *sysroot,
+                                                OstreeDeployment *booted_deployment)
+{
+  g_autoptr(GFile) root = g_file_new_for_path ("/");
+  g_autoptr(GFile) refspec_to_upgrade_file = g_file_new_for_path (REFSPEC_TO_UPGRADE_ON_PATH);
+  g_autoptr(GFile) deployment_directory = ostree_sysroot_get_deployment_directory (sysroot,
+                                                                                   booted_deployment);
+  g_autofree gchar *relative_path = g_file_get_relative_path (root, refspec_to_upgrade_file);
+  return g_file_get_child (deployment_directory, relative_path);
+}
+
+gboolean
+get_refspec_to_upgrade_on (gchar  **refspec_to_upgrade_on,
+                           gchar  **remote_to_upgrade_on,
+                           gchar  **ref_to_upgrade_on,
+                           GError **error)
+{
+  g_autoptr(GFile) refspec_to_upgrade_on_file =  NULL;
+  g_autofree gchar *refspec = NULL;
+  g_autofree gchar *remote = NULL;
+  g_autofree gchar *ref = NULL;
+  g_autoptr(OstreeSysroot) sysroot = ostree_sysroot_new_default ();
+  g_autoptr(OstreeDeployment) booted_deployment = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  if (!ostree_sysroot_load (sysroot, NULL, error))
+    return FALSE;
+
+  booted_deployment = eos_updater_get_booted_deployment_from_loaded_sysroot (sysroot,
+                                                                             error);
+
+  if (booted_deployment == NULL)
+    return FALSE;
+
+  refspec_to_upgrade_on_file = get_refspec_to_upgrade_on_file_from_deployment (sysroot,
+                                                                               booted_deployment);
+
+  if (!g_file_load_contents (refspec_to_upgrade_on_file,
+                             NULL,
+                             &refspec,
+                             NULL,
+                             NULL,
+                             &local_error))
+    {
+      if (!g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return FALSE;
+        }
+
+      /* Couldn't read the refspec file. Just use the booted refspec */
+      return get_booted_refspec (booted_deployment,
+                                 refspec_to_upgrade_on,
+                                 remote_to_upgrade_on,
+                                 ref_to_upgrade_on,
+                                 error);
+    }
+
+  if (!ostree_parse_refspec (refspec, &remote, &ref, error))
```

Comment:
Might want to `g_strstrip (refspec)` first, for robustness. (If someone edits the `upgrade-refspec` file with a text editor, it will probably automatically append a `\n`.)